### PR TITLE
Introduce low shuffle merge.

### LINKED ISSUE #10905: [FEA] Introduce `GpuLowShuffleMergeCommand`.

### PR DESCRIPTION
**Pull request by liurenjie1024**
Close #10905 .
This pr is the first one to introduces low shuffle merge optimization to speed up merge. Currently we only support databricks  13.3, we will add support more versions once this pr gets merged.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuReadFileFormatWithMetrics.scala [hunk 1]
```diff
@@ -33,6 +33,7 @@ import org.apache.spark.sql.types.StructType
 import org.apache.spark.util.SerializableConfiguration
 
 trait GpuReadFileFormatWithMetrics extends FileFormat {
+
```
Comment (razajafri) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuReadFileFormatWithMetrics.scala:36]: unnecessary change

File: delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/shims/MergeIntoCommandMetaShim.scala [hunk 2]
```diff
@@ -17,9 +17,9 @@
 package com.nvidia.spark.rapids.delta.shims
```
Comment (razajafri) [inline delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/shims/MergeIntoCommandMetaShim.scala:17]: copyrights

File: delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala [hunk 3]
```diff
@@ -29,7 +29,8 @@ import org.apache.spark.sql.types.StructType
 case class GpuDeltaParquetFileFormat(
```
Comment (razajafri) [inline delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala:29]: Update copyrights

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaSQLConf.scala [hunk 4]
```diff
@@ -67,6 +67,13 @@ trait RapidsDeltaSQLConf {
       .doc("Maximum amount of data for auto compaction.")
       .bytesConf(ByteUnit.BYTE)
       .createWithDefaultString("20GB")
+
```
Comment (razajafri) [inline delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaSQLConf.scala:70]: Update copyrights

File: sql-plugin/src/main/spark341db/scala/org/apache/spark/sql/execution/rapids/shims/FilePartitionShims.scala [hunk 5]
```diff
@@ -32,4 +32,12 @@ object FilePartitionShims extends SplitFiles {
       }
     }
   }
+
```
Comment (razajafri) [inline sql-plugin/src/main/spark341db/scala/org/apache/spark/sql/execution/rapids/shims/FilePartitionShims.scala:35]: Update Copyrights

**Review by razajafri — COMMENTED** [review]
Do we need to add tests? EIther unit or integration tests?

File: delta-lake/common/src/main/delta-io/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala [hunk 6]
```diff
@@ -33,6 +33,7 @@ import org.apache.spark.util.SerializableConfiguration
 trait GpuDeltaParquetFileFormat extends GpuReadParquetFileFormat {
   val columnMappingMode: DeltaColumnMappingMode
   val referenceSchema: StructType
+  val disableFilterPushdown: Boolean = false
```
Comment (jlowe) [inline delta-lake/common/src/main/delta-io/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala:36]: Seems unrelated to low shuffle merge?

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala [hunk 7]
```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2020-2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala:2]:
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.
```

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object RapidsRepartitionByFilePath {
+  val ROW_ID_COL = "_metadata_row_id"
+  val ROW_ID_FIELD = StructField(ROW_ID_COL, LongType, nullable = false)
+  val FILE_PATH_COL = "_metadata_file_path"
+  val FILE_PATH_FIELD = StructField(FILE_PATH_COL, StringType, nullable = false)
+}
+
+case class RapidsRepartitionByFilePath(optNumPartition: Option[Int],
```
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala:39]: Comments on these classes and what they're doing would help.  Somewhere there needs to be an overview of how these custom operators plugin into the low shuffle merge algorithm.  I haven't seen any comments explaining why we're partitioning on the file path as part of the algorithm, for example.

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+case class GpuRapidsRepartitionByFilePathExec(optNumPartition: Option[Int],
+    gpuPartitionIdExpr: GpuExpression,
+    child: SparkPlan) extends UnaryExecNode with GpuExec {
+  private lazy val numPartitions = optNumPartition.getOrElse(conf.numShufflePartitions)
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = {
+    child.output.find(_.name == FILE_PATH_FIELD.name) match {
+      case Some(attr) => HashPartitioning(Seq(attr), numPartitions)
```
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala:79]: So this new node is likely going to inject an extra shuffle because of this partitioning.  Intentional?

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    s"${getClass.getCanonicalName} does not support row-based execution")
+
+  override protected def internalDoExecuteColumnar(): RDD[ColumnarBatch] = {
+    val rdd = child.executeColumnar()
+    if (numPartitions == 1 && rdd.getNumPartitions < 1) {
+      // Make sure we don't output an RDD with 0 partitions, when claiming that we have a
+      // `SinglePartition`.
+      new GpuCoalesceExec.EmptyRDDWithPartitions(sparkContext, numPartitions)
+    } else {
+      rdd.coalesce(numPartitions,
```
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsRepartitionByFilePath.scala:98]: What if numPartitions > rdd.getNumPartitions?  Without a shuffle, this can't spread to a wider number of partitions.

File: delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala [hunk 11]
```diff
@@ -0,0 +1,1050 @@
...
+import org.apache.spark.sql.types.{BooleanType, LongType, StringType, StructField, StructType}
+
+/**
+ * GPU version of Delta Lake's low shuffle merge implementation.
+ *
+ * Performs a merge of a source query/table into a Delta table.
+ *
+ * Issues an error message when the ON search_condition of the MERGE statement can match
+ * a single row from the target table with multiple rows of the source table-reference.
+ * Different from the original implementation, it optimized writing unmodified target rows.
```
Comment (jlowe) [inline delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala:66]: This sentence implies we're going to rewrite the entire target table even if there are target files that would otherwise be untouched.

File: delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala [hunk 12]
```diff
@@ -0,0 +1,1050 @@
...
+  /**
+   * Repartitions the output DataFrame by the partition columns if table is partitioned
+   * and `merge.repartitionBeforeWrite.enabled` is set to true.
+   */
+  protected def repartitionIfNeeded(
+      spark: SparkSession,
+      df: DataFrame,
+      partitionColumns: Seq[String]): DataFrame = {
+    if (partitionColumns.nonEmpty && spark.conf.get(DeltaSQLConf.MERGE_REPARTITION_BEFORE_WRITE)) {
+      df.repartition(partitionColumns.map(col): _*)
```
Comment (jlowe) [inline delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala:290]: This logic may need to change to consider optimized write, see #10417

**Review by jlowe — COMMENTED** [review]
Have not finished the review yet, but here are some early comments.  Like @razajafri said, there needs to be tests for this along with benchmarking to show the performance vs. baseline CPU and GPU without low shuffle vs GPU with low shuffle in various setups (e.g.: lots of rows updating, very few rows updating, etc.)

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaSQLConf.scala (hunk 4 above)
Comment (thirtiseven) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaSQLConf.scala:70]: We have [pre commit hooks](https://github.com/NVIDIA/spark-rapids/blob/branch-24.06/CONTRIBUTING.md#pre-commit-hooks) to update copyrights automatically.

File: delta-lake/common/src/main/delta-io/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, NVIDIA CORPORATION.
+ * Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/common/src/main/delta-io/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala:2]: Revert this file.

File: delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/DeltaProviderImplBase.scala
Comment (jlowe) [inline delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/DeltaProviderImplBase.scala:1]: 2024 copyrights

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala [hunk 14]
```diff
@@ -0,0 +1,1088 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala:2]:
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.
```

File: delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala [hunk 15]
```diff
@@ -0,0 +1,1083 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala:2]:
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.
```

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala [hunk 16]
```diff
@@ -0,0 +1,1088 @@
...
+      val baseTargetDF = buildTargetDFWithFiles(filesToRewrite)
+
+      // Here we enforce that the targetDF to read each file per partition. This is to ensure
+      // that the `_metadata_row_id` column is correctly populated for each row in the targetDF.
+      val newPlan = baseTargetDF.queryExecution.analyzed.transform {
+        case r@LogicalRelation(fs: HadoopFsRelation, _, _, _) =>
+          val newFs = fs.copy(options = fs.options +
+            (RapidsConf.FORCE_ONE_FILE_PER_PARITION.key -> "true"))(spark)
+          r.copy(relation = newFs)
+      }
```
Comment (jlowe) [inline delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala:648]: This seems like a potential performance problem we will want to address at some point.  We often combine files to get larger data sizes for the GPU to consume, and this will prevent that.  It doesn't seem strictly necessary if we change the monotonically increasing ID to be a true row ID, something that should be relatively straightforward to implement.  Probably worth a followup issue to track.

File: delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala (hunk 12 above)
Comment (jlowe) [reply delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala:290]: That's fine, but probably worth putting a TODO and linking to the issue since we know we're doing it wrong here.

File: delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, NVIDIA CORPORATION.
+ * Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormat.scala:2]: Revert this file.

File: delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/shims/MergeIntoCommandMetaShim.scala [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, NVIDIA CORPORATION.
+ * Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline delta-lake/delta-spark341db/src/main/scala/com/nvidia/spark/rapids/delta/shims/MergeIntoCommandMetaShim.scala:2]: Revert this file.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py [hunk 19]
```diff
@@ -0,0 +1,211 @@
+# Copyright (c) 2023-2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:1]:
```suggestion
# Copyright (c) 2024, NVIDIA CORPORATION.
```

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+# Databricks changes the number of files being written, so we cannot compare logs
+num_slices_to_test = [10] if is_databricks_runtime() else [1, 10]
+
+delta_merge_enabled_conf = copy_and_update(delta_writes_enabled_conf,
+                                           {
+                            "spark.rapids.sql.command.MergeIntoCommand": "true",
+                            "spark.rapids.sql.command.MergeIntoCommandEdge": "true",
+                            "spark.rapids.sql.delta.lowShuffleMerge.enabled": "true"
+                                            })
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:34]: Nit: The indentation is really weird here, can't figure out what rule it's trying to follow.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+    src_table = spark_tmp_table_factory.get()
+    def setup_tables(spark):
+        setup_delta_dest_tables(spark, data_path, dest_table_func, use_cdf, partition_columns)
+        src_table_func(spark).createOrReplaceTempView(src_table)
+    def do_merge(spark, path):
+        dest_table = spark_tmp_table_factory.get()
+        read_delta_path(spark, path).createOrReplaceTempView(dest_table)
+        return spark.sql(merge_sql.format(src_table=src_table, dest_table=dest_table)).collect()
+    with_cpu_session(setup_tables)
+    check_func(data_path, do_merge)
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:53]: This and other stuff seems identical to the versions in delta_lake_merge_test?  Not sure this should be a full-blown separate test suite from delta_lake_merge_test, especially with low-shuffle being on by default.  That means delta_lake_merge_test is actually a low shuffle merge test, since it's on by default, and we're not testing when it's off.  Better to parameterize the existing merge tests in delta_lake_merge_test with low- and non-low- confs than build a separate suite here.  Confs for parameterization can be based on the platform, so we avoid trying to test low-shuffle merge on platforms that do not support it.

File: jenkins/databricks/build.sh [hunk 22]
```diff
@@ -165,5 +165,5 @@ if [[ "$WITH_DEFAULT_UPSTREAM_SHIM" != "0" ]]; then
         -Dincluded_buildvers=$UPSTREAM_BUILDVER,$BUILDVER
 fi
 
-cd /home/ubuntu
-tar -zcf spark-rapids-built.tgz spark-rapids
+#cd /home/ubuntu
+#tar -zcf spark-rapids-built.tgz spark-rapids
```
Comment (jlowe) [inline jenkins/databricks/build.sh:169]: Revert this file

File: jenkins/databricks/test.sh
Comment (jlowe) [inline jenkins/databricks/test.sh:1]: Debugging code that was accidentally pushed to the PR.  Revert this file.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuReadFileFormatWithMetrics.scala [hunk 23]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2023, NVIDIA CORPORATION.
+ * Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuReadFileFormatWithMetrics.scala:2]: Revert this file.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 24]
```diff
@@ -2188,6 +2188,18 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
       .integerConf
       .createWithDefault(1024)
 
+  val FORCE_ONE_FILE_PER_PARITION = conf("spark.rapids.sql.scan.forceOneFilePerPartition")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2191]:
```suggestion
  val FORCE_ONE_FILE_PER_PARTITION = conf("spark.rapids.sql.scan.forceOneFilePerPartition")
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala [hunk 25]
```diff
@@ -554,16 +557,21 @@ case class GpuFileSourceScanExec(
       readFile: Option[(PartitionedFile) => Iterator[InternalRow]],
       selectedPartitions: Array[PartitionDirectory],
       fsRelation: HadoopFsRelation): RDD[InternalRow] = {
-    val openCostInBytes = fsRelation.sparkSession.sessionState.conf.filesOpenCostInBytes
-    val maxSplitBytes =
-      FilePartition.maxSplitBytes(fsRelation.sparkSession, selectedPartitions)
-    logInfo(s"Planning scan with bin packing, max size: $maxSplitBytes bytes, " +
-      s"open cost is considered as scanning $openCostInBytes bytes.")
+    val partitions = if (forceOneFilePerPartition) {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala:560]: This is not the only place where we perform partition setup for scans.  This is also done in GpuBatchScanExec, Iceberg code, etc.  Config needs more detailed comments if it's not supposed to be universally applied to GPU scans.

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala [hunk 26]
```diff
@@ -0,0 +1,1088 @@
...
+      // Generate a new target dataframe that has same output attributes exprIds as the target plan.
+      // This allows us to apply the existing resolved update/insert expressions.
+      val baseTargetDF = buildTargetDFWithFiles(filesToRewrite)
+
+      // Here we enforce that the targetDF to read each file per partition. This is to ensure
+      // that the `_metadata_row_id` column is correctly populated for each row in the targetDF.
+      val newPlan = baseTargetDF.queryExecution.analyzed.transform {
+        case r@LogicalRelation(fs: HadoopFsRelation, _, _, _) =>
+          val newFs = fs.copy(options = fs.options +
+            (RapidsConf.FORCE_ONE_FILE_PER_PARITION.key -> "true"))(spark)
```
Comment (jlowe) [inline delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala:646]: What if the scan does not run on the GPU?  The user can explicitly disable the scan via configs, or there may be some datatype that we cannot load but otherwise can "ride along" in dataframes.  Seems like there needs to be additional tagging logic to catch this and fallback or otherwise handle it properly.  Also needs tests for this corner case.

File: sql-plugin/src/main/spark341db/scala/org/apache/spark/sql/execution/rapids/shims/FilePartitionShims.scala [hunk 27]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, NVIDIA CORPORATION.
+ * Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline sql-plugin/src/main/spark341db/scala/org/apache/spark/sql/execution/rapids/shims/FilePartitionShims.scala:2]:
```suggestion
 * Copyright (c) 2023-2024, NVIDIA CORPORATION.
```

**Review by jlowe — COMMENTED** [review]
Still need performance numbers for various setups and a tracking issue for porting the code to other platforms.

Also note the [user documentation for Delta Lake support](https://docs.nvidia.com/spark-rapids/user-guide/24.04/additional-functionality/delta-lake-support.html) will need to be updated to describe this new feature after it's merged.

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala (hunk 16 above)
Comment (jlowe) [reply delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala]: Then why not have the scan partitioning match that behavior rather than forcing one per file?  We could still group files in a task if the filepaths hash to the same partition, keeping in mind how many partitions we're trying to target later.  Forcing a single file per task can lead to very poor performance when the files are small, which can be the case when the table is partitioned.

Comment (jlowe) [reply delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala]: I'm fine with a followup issue to track, but it's probably not something we want to put off too long.  I'm also not sure we really need a true row ID separate from the one Delta Lake is already using when it finds touched files.  It uses a generated row ID column implemented via MonotonicallyIncreasingID.  That ID generates a unique ID for each row across the dataframe, because it has a partition ID in the top 31 bits and a within-partition row number in the bottom 33 bits, meaning the ID is going to be unique across the entire DataFrame.

So I think we only need to consider the hashed file path and split size when trying to group into the desired number of partitions, and we would be able to combine some smaller files.

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaUtils.scala [hunk 28]
```diff
@@ -99,4 +101,35 @@ object RapidsDeltaUtils {
...
+          .filterNot(f => ScanExecShims.supportByGpuFileSourceScanExec(f.dataType))
+
+        if (unsupportedFields.nonEmpty) {
+          logWarning(
+            s"""Delta lake low shuffle merge can't be enabled because following fields in
+               |target table could be supported by GpuFileSourceScanExec:
+               |${unsupportedFields.mkString("Array(", ", ", ")")}"""
+              .stripMargin)
+          false
+        } else {
```
[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaUtils.scala]: Honestly I'm not thrilled with the extra new exec node, and it's dangerous as I mentioned here to have it, because it makes blind assumptions about its input that may not be correct if the upstream stuff didn't do it's job as expected.

I think it would be safer and cleaner if GpuFileSourceScanExec directly advertised it's output as being hash partitioned by the file path when it's in this special mode.  In other words, put the logic of the new exec in GpuFileSourceScanExec.  Then we're safely covered if the scan falls back to the CPU, because Spark itself will inject the necessary shuffle for the join rather than performing an invalid join and producing bad output.  It also removes the extra exec from the plan which may confuse users as to what its doing (as it has no CPU correlation).

[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/RapidsDeltaUtils.scala]: Ah right, I was thinking these would be injected between GpuOverrides and GpuOverrideTransitions, but that's only columnar transitions not shuffle exchanges.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py (hunk 21 above)
Comment (jlowe) [reply integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:51]: Is this still going to get fixed?  There's needless duplication of lots of test code here that I don't see why we need to postpone.  It's trivial to fix.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 29]
```diff
@@ -2194,6 +2194,20 @@ val SHUFFLE_COMPRESSION_LZ4_CHUNK_SIZE = conf("spark.rapids.shuffle.compression.
       .integerConf
       .createWithDefault(1024)
 
+  val FORCE_ONE_FILE_PER_PARITION =
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:2197]: If we're also requiring the per-file reader to be enabled, then another way to accomplish this is to disable splitting on DeltaParquetFileFormat, since per-file will force file-at-a-time in a task and unsplittable input will prevent files from being shared across tasks.  Not sure if there's a clean way to control the splittability of DeltaParquetFileFormat when merging, but if so, would be a cleaner solution that this hidden config.

File: jenkins/spark-premerge-build.sh [hunk 30]
```diff
@@ -206,7 +206,7 @@ ci_scala213() {
     cd .. # Run integration tests in the project root dir to leverage test cases and resource files
     export TEST_TAGS="not premerge_ci_1"
     export TEST_TYPE="pre-commit"
-    export TEST_PARALLEL=5
+    export TEST_PARALLEL=4
```
[the remark this answers is not in the document]
Comment (jlowe) [reply jenkins/spark-premerge-build.sh:209]: This doesn't seem like a change that should be made as part of Delta Lake low shuffle merge but rather as a separate PR, especially if you can get it to fail without your low shuffle merge changes.  cc: @NvTimLiu for visibility.

[the remark this answers is not in the document]
Comment (jlowe) [reply jenkins/spark-premerge-build.sh:209]: What I'm saying above is that this change is not really related to this PR.  It's a significant change in CI scripts that will affect performance of CI, since we'll run fewer tests in parallel.  That's why I think this should be a separate change, not hidden in a large PR as a side-effect.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+def test_delta_merge_when_fallback(spark_tmp_path, spark_tmp_table_factory, use_cdf,
+                                       num_slices):
+    # Need to eliminate duplicate keys in the source table otherwise update semantics are ambiguous
+    src_table_func = lambda spark: two_col_df(spark, int_gen, string_gen, num_slices=num_slices).groupBy("a").agg(f.max("b").alias("b"))
+    dest_table_func = lambda spark: two_col_df(spark, int_gen, string_gen, seed=1, num_slices=num_slices)
+    merge_sql = "MERGE INTO {dest_table} USING {src_table} ON {dest_table}.a == {src_table}.a" \
+                " WHEN MATCHED THEN UPDATE SET * WHEN NOT MATCHED THEN INSERT *"
+
+    conf = copy_and_update(delta_merge_enabled_conf,
+                           {"spark.rapids.sql.exec.RapidsRepartitionByFilePathExec": "false"})
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:47]: This conflates the original WHEN fallback test with another test checking for CPU fallback of scan or partition.  They need to be separate tests, otherwise we're not actually testing if we fallback to the CPU for a normal WHEN clause as expected, only testing that when a, probably never set in practice, unrelated config is also set.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+from pyspark.sql.types import *
+from spark_session import *
+
+delta_merge_enabled_conf = copy_and_update(delta_writes_enabled_conf,
+                                           {"spark.rapids.sql.command.MergeIntoCommand": "true",
+                            "spark.rapids.sql.command.MergeIntoCommandEdge": "true",
+                            "spark.rapids.sql.delta.lowShuffleMerge.enabled": "true",
+                            "spark.rapids.sql.format.parquet.reader.type": "PERFILE"})
+
+@allow_non_gpu("RapidsRepartitionByFilePathExec", *delta_meta_allow)
```
Comment (jlowe) [inline integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:29]: How does RapidsRepartionByFilePathExec work if it isn't converted to GpuRapidsRepartitionByFilePathExec?  The former has a doExecute method that will throw, so seems like this would just crash in practice if it wasn't converted to the GPU version.

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala [hunk 33]
```diff
@@ -0,0 +1,1166 @@
...
+
+    private def checkGpuFileSourceScanExecOverride(input: DataFrame): Boolean = {
+        val executedPlan = input.queryExecution
+          .executedPlan
+
+        var pushed = false
+        GpuOverrides().applyWithContext(executedPlan.clone(), Some(
+          """Detecting GpuFileSourceScanExec could be pushed down""".stripMargin)).foreachUp {
+          case GpuRapidsRepartitionByFilePathExec(_, _,
+          GpuProjectExec(_, _: GpuFileSourceScanExec)) =>
```
Comment (jlowe) [inline delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala:1158]: This seems very specific and fragile.  Aren't there usually filters inserted by Spark immediately after a scan to remove nulls before a join, for example?  We don't actually care if there's a row-to-columnar or columnar-to-row in here, we just need there to be a GpuFileSourceScanExec underneath the GpuRapidsRepartitionByFilePathExec without partitioning changes in-between, correct?

File: delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala [hunk 34]
```diff
@@ -0,0 +1,1169 @@
...
+        val partitionedTargetDF = addRepartitionByFilePath(baseTargetDF)
+        val df = makeUnmodifiedTargetDF(partitionedTargetDF, modifiedTargetRows)
+
+        // The correctness of RapidsRepartitionByFilePath relies on the fact we can enable
+        // FORCE_ONE_FILE_PER_PARITION in file scan, which currently only works for
+        // GpuFileSourceScanExec, so we have to rollback to original input if for some reason we
+        // can use GpuFileSourceScanExec.
+        val gpuFileScanOverride = checkGpuFileSourceScanExecOverride(df)
+
+        if (!gpuFileScanOverride) {
```
[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala:954]: The test falls back the new custom repartition exec (which I'm confused how this works since the CPU version of this exec does not allow execution).  Seems like the test should test what is actually expected to happen in practice (i.e.: the scan falling back) rather than the custom exec that no user knows about falling back.

File: integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py (hunk 32 above)
Comment (jlowe) [reply integration_tests/src/main/python/delta_lake_low_shuffle_merge_test.py:29]: > It's not actually executed

Curious how is it not executed in practice?  I thought it would be a child node of the join, as that's the whole point of it to repartition without actually shuffling for the join.  The join is going to try to execute it's child nodes to perform the join, which seems fatal if the CPU RapidsRepartitionByFilePathExec is in the plan.  The GPU version of the node has an execute (columnar) implementation, so that makes it even more confusing how the CPU version does not need one.

**Review by jlowe — COMMENTED** [review]
Thanks for all the updates, @liurenjie1024!  This is getting close.  Would be good to file the followup issues, ideally pointing to them with TODO's in the code.  Also need performance numbers as mentioned before.
